packages/react-dom-bindings/src/client/ReactFiberConfigDOM.js

```diff
@@ -1534,6 +1534,206 @@ export function subscribeToGestureDirection(
+  ): void,
+  removeEventListener(
+    type: string,
+    listener: EventListener,
+    optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+  ): void,
+  focus(): void,
+};
+
+function FragmentInstancePseudoElement(
+  this: FragmentInstance,
+  fragmentFiber: Fiber,
+) {
+  this._fragmentFiber = fragmentFiber;
+}
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.addEventListener = function (
+  this: FragmentInstance,
+  type: string,
+  listener: EventListener,
+  optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+): void {
+  if (!this._eventListeners) {
+    this._eventListeners = [];
+  }
+
+  const listeners = this._eventListeners;
+  // Element.addEventListener will only apply uniquely new event listeners by default. Since we
+  // need to collect the listeners to apply to appended children, we track them ourselves and use
+  // custom equality check for the options.
+  const isNewEventListener =
+    indexOfEventListener(listeners, {
+      type,
+      listener,
+      optionsOrUseCapture,
+    }) === -1;
+  if (isNewEventListener) {
+    listeners.push({type, listener, optionsOrUseCapture});
+    const children = getFragmentInstanceChildren(this);
+    children.forEach(child => {
+      child.addEventListener(type, listener, optionsOrUseCapture);
+    });
+  }
+  this._eventListeners = listeners;
+};
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.removeEventListener = function (
+  this: FragmentInstance,
+  type: string,
+  listener: EventListener,
+  optionsOrUseCapture?: EventListenerOptionsOrUseCapture,
+): void {
+  const listeners = this._eventListeners;
+  if (listeners === undefined || listeners.length === 0) {
+    return;
+  }
+
+  const children = getFragmentInstanceChildren(this);
+  children.forEach(child => {
+    child.removeEventListener(type, listener, optionsOrUseCapture);
+  });
+  const index = indexOfEventListener(listeners, {
+    type,
+    listener,
+    optionsOrUseCapture,
+  });
+  this._eventListeners = listeners.splice(index, 1);
+};
+// $FlowFixMe[prop-missing]
+FragmentInstancePseudoElement.prototype.focus = function (
+  this: FragmentInstance,
+) {
+  const children = getFragmentInstanceChildren(this);
+  const iterator = children.values();
+  let child = iterator.next();
+  while (!child.done) {
+    if (setFocusIfFocusable(child.value)) {
+      break;
+    }
+    child = iterator.next();
+  }
+};
+
+function getFragmentInstanceChildren(
+  fragmentInstance: FragmentInstance,
+): Set<Element> {
+  const children = new Set<Element>();
+  recursivelyGetFragmentInstanceChildren(
+    fragmentInstance._fragmentFiber.child,
+    children,
+  );
+  return children;
+}
+
+function recursivelyGetFragmentInstanceChildren(
+  child: Fiber | null,
+  childElements: Set<Element>,
+): void {
+  if (child === null) {
+    return;
+  }
+
+  if (child.tag === HostComponent) {
+    childElements.add(child.stateNode);
+  }
+
+  if (child.sibling !== null) {
```

Comment:
It's typically better to do this with a loop since otherwise the recursion can get very deep and lead to stack overflow.

Since React does use a stack we do require a tree to be more shallow than required for a typical recursion but we don't require it to also handle sibling depth. You can have long list of items in deep trees.

Reply:
This also leads to a sort of breadth first search which I don't think is necessarily the intended order of insertion.

```
<Fragment>
  <Component><div id="a" /></Component>
  <div id="b" />
</Fragment>
```

This should have "a" first and then "b" but the current algorithm would visit "b" first.

Look at how we do other traversals where we do siblings in a loop and then recurse each child in that loop. Depth first.

Reply:
Updated to be depth first and loop the siblings.